0005-kzen-two-party-ecdsa.md

```diff
@@ -0,0 +1,139 @@
+But in a more recent paper on multi-party ECDSA Lindell mentioned a more efficient protocol to make the proof: [https://eprint.iacr.org/2018/987.pdf](https://eprint.iacr.org/2018/987.pdf).
+See the section 6.2.3, which explains how to modify the proof in [https://eprint.iacr.org/2018/057.pdf](https://eprint.iacr.org/2018/057.pdf) to simply prove `N` is generated fairly.
+
+Roughly speaking, the strategy is to prove that `N` is *square free* and to show that `f(x) = x ^ N (mod N)` is a permutation which proves that `N = pq`.
+
+#### Implementation notes
+
+Since it's a public coin protocol, you can make an non-interactive version of the proof using the [Fiat-Shamir transformation](https://en.wikipedia.org/wiki/Fiat%E2%80%93Shamir_heuristic).
+They've implemented a non-interactive version of it here: https://github.com/KZen-networks/zk-paillier/blob/a7f3907fb2b3464f12b55e3ad2f50405db54f36a/src/zkproofs/correct_key_ni.rs (which uses the interactive proof module but supplies the challenge as a hash of the initial commitment).
+
+### Proving a Paillier encryption is a private key of a EC public key
+
+In order to produce two party aggregated ECDSA signatures on a joint Secp256k1 public key we need one party (P1) their Paillier encrypted private key over to P2.
+But in order for P2 to know that he's signing under their joint public key, he has to know that this encrypted private key is the right one (without learning what the private key is).
+
+The proof for this was developed in paper. It's a little tricky to get your head around.
+It works by the P2 doing some homomorphic operations on the encrypted private key and sending the result to P1.
+P1 then decrypts it and figures out the corresponding public key.
+
+The weird thing is that P1 doesn't actually use his knowledge of what the private key is in order to produce the proof that the cipher text doesn't contain the private key.
+He simply uses his ability to decrypt the Paillier private key!
+
+#### Implementation notes
+
+This proof isn't in its own module but the code for it hangs around with the rest of the protocol:
+
+https://github.com/KZen-networks/multi-party-ecdsa/blob/e5a741bf8dd756b650b35ef8d65f6cecbd4f196a/src/protocols/two_party_ecdsa/lindell_2017/party_one.rs
+
+It looks like this protocol cannot be made non-interactive. I think it's the only one that couldn't be so it accounts for most of the rounds of communication.
+It has four rounds of communication.
+
+### Range proof
+
+In order for the previous proof to actually prove the statement you have to couple it with a range proof which proves that the encrypted private key is in the curve order (i.e. is a valid private key).
+The poof chosen was originally from [https://www.iacr.org/archive/eurocrypt2000/1807/18070437-new.pdf](https://www.iacr.org/archive/eurocrypt2000/1807/18070437-new.pdf) but I found it was easier to understand in Lindell's paper anyway (see Appendix A).
```

Comment:
Click-able references (for the "Lindell's paper" part) would be really good here (plus IEEE citation and a list in the end).